### PR TITLE
gh-127740: For odd-length input to bytes.fromhex(...) change the error message to  ValueError: fromhex() arg must be of even length

### PR DESCRIPTION
**Pull request by srinivasreddy**
* Issue: gh-127740

### CODE REVIEW

**Review by hauntsaninja — COMMENTED** [review]
Can we avoid the upfront loop you're using to check this eagerly? E.g. check invalid_char vs hexlen in the error clause?

**Comment by srinivasreddy** [comment]
@hauntsaninja Please review again?

File: Objects/bytesobject.c [hunk 1]
```diff
@@ -2526,21 +2526,34 @@ _PyBytes_FromHex(PyObject *string, int use_bytearray)
     end = str + hexlen;
     while (str < end) {
         /* skip over spaces in the input */
-        if (Py_ISSPACE(*str)) {
-            do {
-                str++;
-            } while (Py_ISSPACE(*str));
-            if (str >= end)
-                break;
+        while (str < end && Py_ISSPACE(*str)) {
```
Comment (picnixz) [inline Objects/bytesobject.c]: Why this change? are performances improved?

**Review by picnixz — COMMENTED** [review]
I think hauntsaninja wanted the error check at the level of the error label, not during the parsing. Since the current code jumps to error in case of an odd-length input, is it possible to check that it's indeed odd-length when creating the error?

File: Objects/bytesobject.c (hunk 1 above)
Comment (srinivasreddy) [reply Objects/bytesobject.c]: I found this logic easier to follow. Not recommended to change? I didn't measure performance.

File: Lib/test/test_bytes.py [hunk 2]
```diff
@@ -459,6 +459,12 @@ def test_fromhex(self):
         self.assertRaises(ValueError, self.type2test.fromhex, '\x00')
         self.assertRaises(ValueError, self.type2test.fromhex, '12   \x00   34')
 
+        # For odd number of character(s)
+        for value in ("a", "a ", " a"," a ", "aaa", "aaa ", " aaa", " aaa ",  " aa a "):
```
Comment (srinivasreddy) [inline Lib/test/test_bytes.py]: What should be the behaviour of `fromhex(" aa a ")`? What message should it throw?

1.fromhex() arg must be of even length
2.non-hexadecimal number found in fromhex() arg at position 

Confused.

Comment (srinivasreddy) [reply Lib/test/test_bytes.py]: Please review again and let me know @picnixz

Comment (picnixz) [reply Lib/test/test_bytes.py]: The error "must be of even length" would be confusing since it *is* of even length, so I'd say "non-hexadecimal number found in fromhex() arg at position" so that users could remove the spaces *and then* they would see that there is an odd number of *valid* characters. 

Since we are leaning towards improving user experience, having a confusing error message isn't really nice (namely, let's not replace a confusing message by another confusing one). 

But we need the opinion of others. What do you prefer for instance?

Comment (srinivasreddy) [reply Lib/test/test_bytes.py]: I agree with your logic.

Comment (hauntsaninja) [reply Lib/test/test_bytes.py]: I think we should change the error message to say "must contain an even number of hexadecimal digits" which will help with the first thing picnixz mentions, since that wording indicates we're no longer talking about string length

I don't have a strong opinion on what error message " aa a " or " aa a a " raises. [It looks like an intentional feature that it disallow pairs with space between them](https://github.com/python/cpython/commit/0b9b9e04830c40c7866b4ba23d6fa1a0aa80f039#diff-58df2535f64e3dcfb0677a7f543924d976fd982b595f8e34902a73f3deae40ebR410), so I'm inclined to keep this example raising "non-hexadecimal number found in fromhex()"

Comment (srinivasreddy) [reply Lib/test/test_bytes.py]: Which one you prefer? Or something else?

1. `s/"fromhex() arg must be of even length"/"must contain an even number of hexadecimal digits"`
2. `s/"fromhex() arg must be of even length"/"fromhex() arg must contain an even number of hexadecimal digits"`

Comment (hauntsaninja) [reply Lib/test/test_bytes.py]: second one was what I was thinking

**Comment by cfbolz** [comment]
Yes, I agree with @picnixz and @hauntsaninja, the new if is done for every character pair now, which makes things less efficient.

I suggest moving the new check into the condition just below. `if (bot >= 16) {` is true if we are at the end of the string (because the null at the end is mapped to 37), so we can do the check in there.

**Comment by cfbolz** [comment]
No, that's not what I meant, sorry for being unclear. *many* invalid byte values are mapped to 37, not just null. So instead, you need to check for your original condition `if (str >= end) {`, but in the same place where you now added the `if (bot == 37){` check.

**Comment by srinivasreddy** [comment]
@cfbolz what do you think now ?

**Comment by cfbolz** [comment]
@srinivasreddy code looks good now, imo. please add a [news entry](https://devguide.python.org/core-developers/committing/#how-to-add-a-news-entry) though.

**Review by cfbolz — APPROVED** [review]
Looks good to me

File: Lib/test/test_bytes.py [hunk 3]
```diff
@@ -459,6 +459,12 @@ def test_fromhex(self):
         self.assertRaises(ValueError, self.type2test.fromhex, '\x00')
         self.assertRaises(ValueError, self.type2test.fromhex, '12   \x00   34')
 
+        # For odd number of character(s)
+        for value in ("a", "aaa", "deadbee"):
```
Comment (hauntsaninja) [inline Lib/test/test_bytes.py:463]: nit: would be nice to have a test case where the string is of even length, e.g.  " aaa" and test cases where we check the error message for " aa a " / " aa a a "

Comment (cfbolz) [reply Lib/test/test_bytes.py:463]: sorry for going ahead and merging @hauntsaninja. I should have waited for more input from you.

Comment (srinivasreddy) [reply Lib/test/test_bytes.py:463]: https://github.com/python/cpython/pull/127818

Comment (hauntsaninja) [reply Lib/test/test_bytes.py:463]: oh all good, was just a nit and commits are cheap :-) thanks for reviewing!

**Comment by Kodiologist** [comment]
That was fast. Thanks.

**Review by Redeyder723 — APPROVED** [review]
